### DISCUSSION TITLE
How does MVIKotlin compares with MVICore

### DISCUSSION THREAD

**Discussion by belemaire** [badoo/MVIKotlin#257] [General]
In the README it is mentioned that `MVIKotlin` is inspired by `MVICore` library.

It would be super nice to elaborate a bit further on this point _(ideally in the README itself)_ to give a better picture of the commonalities and differences with `MVICore`, briefly explaining why MVIKotlin was made in the first place _(answering the "why yet another MVI Framework")_. 

For newcomers to MVI, the differences/commonalities between these two frameworks are not "screaming" at first glance _(even though I could identify a few principal ones quicky)_, making it difficult to compare the two _(unless by digging into both documentations and implementations)_.

**Comment by arkivanov** [comment]
Hello @belemaire. Because a similar question was already asked in the  the Kotlinlang [#mvikotlin](https://kotlinlang.slack.com/archives/C01403U1ZGW) channel, I will just copy my answer from there.

Both libraries provide similar API for writing Stores (Features in MVICore). However there are some key differences:

- In MVICore it is not possible to read the current (updated) state from inside the Feature. So if you need to check the new state after it is changed, you have to use `PostProcessor`. This gives MVICore some additional purity, which may positively impact maintainability. In MVIKotlin you have `getState: () -> State` supplier that always returns the current state. This reduces boiler plate code, sometimes significantly, and so there is no `PostProcessor` required. So purity vs boilerplate.
- In MVICore if you want to publish some `News` (aka side effects), you have to use `NewsPublisher`. This is again due to that purity. In MVIKotlin you can publish `Labels` (same thing as `News` in MVICore) directly from `Executor`, and so no `NewsPublisher` is required. Again purity vs boilerplate code.
- MVICore is developed and heavily used by Badoo/Bumble, which is quite a big company. So it is proved to work well in production. MVIKotlin is developed by me in my spare time, however there are known users/companies as well using MVIKotlin in production.
- If you are planning to use Kotlin Multipaltform, then at the moment you can't use MVICore in common code, because it depends on RxJava2. MVIKotlin is multiplatform library, and supports Reaktive and Coroutines.
- The way how you create Stores/Features is also different. In MVICore you are encouraged to extend `BaseFeature` class or one of its descendants, and pass components to super. In MVIKotlin you usually use the `StoreFactory` interface where you pass things, and it creates an implementation for you.
- Both libraries provide logging and time travel functionality. However their implementation is also different. In MVICore there are middlewares, logging and time travel are middlewares. There is a global middleware registry, which is used by the library internally. In MVIKotlin there are separate Store implementation for logging and time travel, and so separate store factories. You can combine factories or even implement your own Store implementations.
- There is also such a thing as `Binder` in both libraries. It makes lifecycle-aware connections between inputs and outputs. However in MVICore the `Binder` also has additional responsibility to wrap consumers into middlewares from the global registry.

**Reply by belemaire** [reply]
Thanks much for your quick reply !
This will also be helpful to other developers having the same question and not -yet- on [#mvikotlin](https://kotlinlang.slack.com/archives/C01403U1ZGW) channel.
